### PR TITLE
[Model] Initial support for LLaVA-NeXT

### PR DESCRIPTION
**Pull request by DarkLight1337**
I have added experimental support for LLaVA-NeXT, with one big caveat: the size of the input image is fixed by the configuration, otherwise the feature size (i.e. number of <image> tokens to duplicate) would vary depending on the runtime input. This prevents us from taking full advantage of the extra resolution. Still, this provides us access to a 34b model which should improve over their 7b and 13b LLaVA-1.5 models.

### Related Contributions

This PR completes part of #3978.

Since this PR depends on the functionalities proposed in #4197 to pass `image_sizes` to LLaVA-NeXT model, it is set to Draft status until that is merged. Afterwards, you should be able to see the diffs that are exclusive to this PR.

To avoid unnecessary resource usage, this branch is frozen (except for critical fixes) until its dependencies have all been merged.

### Features

1. Experimental support for LLaVA-NeXT (also known as: LLaVA-1.6, [LLaVA-1.5-HD](https://static.hliu.cc/files/llava/improved_llava.pdf))
    - Added `LlavaNextForConditionalGeneration` to the list of supported architectures. (Tested with [`llava-hf/llava-v1.6-34b-hf`](https://huggingface.co/llava-hf/llava-v1.6-34b-hf))
    - **Limitation:** The input image is resized to a static `image_input_shape` (NCHW format, specified in the configuration) *before* passing it to the model; otherwise, the number of `<image>` input tokens required in the text prompt (equal to `image_feature_size`) would vary at runtime depending on the original size of the input image. The following table shows the `image_feature_size` which you need to specify in the configuration for each `image_input_shape`:

        | **Width (→)<br>Height (↓)** | **336** | **448** | **560** | **672** | **784** | **896** | **1008** | **1120** | **1232** | **1344** |
        | ------------ | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
        | **336**  | 1176 | 1368 | 1560 | 1752 | 1944 | 2136 | 2328 | 2182 | 2036 | 1890 |
        | **448**  | 1376 | 2928 | 2438 | 2144 | 1948 | 1752 | 1896 | 2040 | 2184 | 2328 |
        | **560**  | 1576 | 2448 | 2928 | 2536 | 2242 | 2046 | 1850 | 1752 | 1848 | 1992 |
        | **672**  | 1776 | 2160 | 2544 | 2928 | 2634 | 2340 | 2144 | 1948 | 1850 | 1752 |
        | **784**  | 1976 | 1968 | 2256 | 2640 | 2928 | 2634 | 2438 | 2242 | 2046 | 1948 |
        | **896**  | 2176 | 1776 | 2064 | 2352 | 2640 | 2928 | 2634 | 2438 | 2242 | 2144 |
        | **1008** | 2376 | 1926 | 1872 | 2160 | 2448 | 2640 | 2928 | 2732 | 2536 | 2340 |
        | **1120** | 2232 | 2076 | 1776 | 1968 | 2256 | 2448 | 2736 | 2928 | 2732 | 2536 |
        | **1232** | 2088 | 2226 | 1876 | 1872 | 2064 | 2256 | 2544 | 2736 | 2928 | 2732 |
        | **1344** | 1944 | 2376 | 2026 | 1776 | 1968 | 2160 | 2352 | 2544 | 2736 | 2928 |
        - For other image sizes, you can attempt to run the model first; the resulting error should tell you the expected `image_feature_size`.

### Future Work

We can overcome the current limitations of static input shape after #5215 has been merged. Once that is addressed, we can openly support this model by adding it to the docs and README. See:

- ~~#5279~~ #5276

### CODE REVIEW

**Comment by jeejeelee** [comment]
Nice work,  any plan to port CLIPVisionModel's code?

**Comment by DarkLight1337** [comment]
> Nice work, any plan to port CLIPVisionModel's code?

@jeejeelee It is outside the scope of this PR; however you are welcome to voice your thoughts in #4194.

**Comment by Aurorana** [comment]
Hello, I am using your LLava-Next for inference acceleration, but your Test part looks a bit complicated. Can you give a simple example of running LLava-Next?

**Comment by DarkLight1337** [comment]
> Hello, I am using your LLava-Next for inference acceleration, but your Test part looks a bit complicated. Can you give a simple example of running LLava-Next?

Please take a look at `docs/source/models/VLM.rst` introduced by this PR.

**Comment by AmazDeng** [comment]
Are there any plans to support the llava-next-video project?

**Comment by DarkLight1337** [comment]
> Are there any plans to support the llava-next-video project?

It's too early to consider that. At least wait until LLaVA-NeXT is officially supported.

**Comment by AmazDeng** [comment]
> LLaVA-NeXT is officially supported

So when will the official support for llava-next be available?

**Comment by DarkLight1337** [comment]
> > LLaVA-NeXT is officially supported
> 
> So when will the official support for llava-next be available?

There is no concrete ETA at the moment. Please check #3978 and #4197 for more information.

**Comment by WanBenLe** [comment]
llava or llava-next will support awq or not?

**Comment by DarkLight1337** [comment]
> llava or llava-next will support awq or not?

We are focusing on getting it to work on default settings first. Support for other features may be considered after the initial release.

**Comment by DarkLight1337** [comment]
Finally done merging. Note that the LLaVA-NeXT tests currently fail due to the need to resize the image beforehand, making the output inconsistent with HuggingFace.

**Comment by NaNillll** [comment]
Hello, thanks for your effort for supporting LLAVA-NEXT on vLLM！

Our team also want to implement this. We use LlavaNextProcessor to pre-process image. We suppose it should be 5*3*336*336 according to paper, and we have get corrcect result on image like 1280*1920.
Howerver, when input small size image such as 336*336, the output of LlavaNextProcessor  is 3*3*336*336. Why? This is confusion and also lead to shape error when cat image feature and image_newline.

So do you know why the output size of LlavaNextProcessor is different? How did you solve this problem? Thank you very much!

**Comment by DarkLight1337** [comment]
Please refer to the Limitation section in the OP.

**Comment by NaNillll** [comment]
> Please refer to the Limitation section in the OP.
> 
> **Edit:** Are you referring to our implementation of the image processor or the one on HuggingFace?
We are using LlavaNextProcessor from HuggingFace. And we have set image_feature_size to 1176 according to your tables

**Comment by DarkLight1337** [comment]
> > Please refer to the Limitation section in the OP.
> > 
> > **Edit:** Are you referring to our implementation of the image processor or the one on HuggingFace?
> 
> We are using LlavaNextProcessor from HuggingFace. And we have set image_feature_size to 1176 according to your tables 

Please note that currently we have to resize the image to the config value *before* it is processed by HF processor.

**Comment by NaNillll** [comment]
> > > Please refer to the Limitation section in the OP.
> > > **Edit:** Are you referring to our implementation of the image processor or the one on HuggingFace?
> > 
> > 
> > We are using LlavaNextProcessor from HuggingFace. And we have set image_feature_size to 1176 according to your tables
> 
> Please note that currently we have to resize the image to the config value _before_ it is processed by HF processor. (As discussed in the Limitation session)

So which config value?  (336, 336)/(672, 672)/(1344, 336)/(336, 1344) in est_llava_next.py or the five size in llava config.json?

**Comment by DarkLight1337** [comment]
> > > > Please refer to the Limitation section in the OP.
> > > > **Edit:** Are you referring to our implementation of the image processor or the one on HuggingFace?
> > > 
> > > 
> > > We are using LlavaNextProcessor from HuggingFace. And we have set image_feature_size to 1176 according to your tables
> > 
> > 
> > Please note that currently we have to resize the image to the config value _before_ it is processed by HF processor. (As discussed in the Limitation session)
> 
> So which config value? (336, 336)/(672, 672)/(1344, 336)/(336, 1344) in test_llava_next.py or the five size in llava config.json?

The one in `test_llava_next.py`.

**Comment by NaNillll** [comment]
> > > > > Please refer to the Limitation section in the OP.
> > > > > **Edit:** Are you referring to our implementation of the image processor or the one on HuggingFace?
> > > > 
> > > > 
> > > > We are using LlavaNextProcessor from HuggingFace. And we have set image_feature_size to 1176 according to your tables
> > > 
> > > 
> > > Please note that currently we have to resize the image to the config value _before_ it is processed by HF processor. (As discussed in the Limitation session)
> > 
> > 
> > So which config value? (336, 336)/(672, 672)/(1344, 336)/(336, 1344) in test_llava_next.py or the five size in llava config.json?
> 
> The one passed to the vLLM config (not HF config), e.g. the ones in `test_llava_next.py`.

Ok I konw. Thank you!

File: vllm/model_executor/models/llava_next.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+                        self.image_newline[:, None, None] \
+                            .expand(*other_patch_embeds.shape[:-1], 1) \
+                            .to(other_patch_embeds.device),
+                    ), dim=-1)
+                    other_patch_embeds = other_patch_embeds \
+                        .flatten(1, 2).transpose(0, 1)
+                else:
+                    other_patch_embeds = other_patch_embeds \
+                        .permute(0, 2, 1, 3, 4).contiguous() \
+                        .flatten(0, 3)
```
Comment (ywang96) [inline vllm/model_executor/models/llava_next.py:219]: Should we add a comment of reference here regarding this?

File: vllm/model_executor/models/llava_next.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+                                 f"Got type: {type(image_sizes)}")
+
+            return LlavaNextImagePixelInputs(
+                type="pixel_values",
+                data=self._validate_image_pixels(pixel_values),
+                image_sizes=self._validate_image_sizes(image_sizes),
+            )
+
+        if expected_input_type == ImageInputType.IMAGE_FEATURES:
+            raise TypeError("Image features are not supported by LLaVA-NeXT")
```
Comment (ywang96) [inline vllm/model_executor/models/llava_next.py:169]: Hmm - I think we should move this into initialization time.

**Review by ywang96 — COMMENTED** [review]
Left some comments - thanks for working on this! @DarkLight1337

File: tests/multimodal/test_processor.py [hunk 3]
```diff
@@ -12,7 +12,7 @@
 @pytest.mark.parametrize("dtype", ["half", "float"])
 def test_clip_image_processor(hf_images, dtype):
     MODEL_NAME = "llava-hf/llava-1.5-7b-hf"
-    IMAGE_HEIGHT = IMAGE_WIDTH = 33
+    IMAGE_HEIGHT = IMAGE_WIDTH = 560
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply tests/multimodal/test_processor.py:15]: I think I originally made a typo (meant to be 336 instead of 33). But I should have made the image larger than that anyway to test whether HF does resizing in the same way.

File: vllm/model_executor/models/llava_next.py (hunk 1 above)
Comment (DarkLight1337) [reply vllm/model_executor/models/llava_next.py:219]: Sure!

File: vllm/model_executor/models/llava_next.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+    if isinstance(image, torch.Tensor):
+        pixel_values = image.to(model_config.dtype)
+        batch_size, _, _, h, w = pixel_values.shape
+        image_sizes = torch.tensor([(w, h) for _ in range(batch_size)])
+
+        return {"pixel_values": pixel_values, "image_sizes": image_sizes}
+
+    # Temporary patch before dynamic number of image tokens is supported
+    _, _, h, w = vlm_config.image_input_shape
+    data.image = image.resize((w, h))
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/llava_next.py:75]: Done!

File: vllm/model_executor/models/llava_next.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+    return MULTIMODAL_REGISTRY._get_plugin_for_data_type(ImagePixelData) \
+        ._default_input_processor(data, model_config, vlm_config)
+
+
+@MULTIMODAL_REGISTRY.register_image_pixel_input(_image_pixel_processor)
+@MULTIMODAL_REGISTRY.register_dummy_data(_get_dummy_image_data)
+class LlavaNextForConditionalGeneration(LlavaForConditionalGeneration):
```
[the remark this answers is not in the document]
Comment (DarkLight1337) [reply vllm/model_executor/models/llava_next.py:83]: Done!

**Comment by DarkLight1337** [comment]
Thanks for the review! I have addressed your comments.

**Review by ywang96 — APPROVED** [review]
LGTM! Thanks for adding this model and I'll merge this after pushing a final addition on the docs.

**Comment by DarkLight1337** [comment]
Oops, turns out that I forgot to copy some functions from LLaVA into LLaVA-NeXT. Drafting a quick fix now.

**Comment by nittaya3** [comment]
Hello how are you @DarkLight1337 @jeejeelee @Aurorana @Aurorana @AmazDeng @AmazDeng @WanBenLe @NaNillll @Don-Chad @ywang96 @pcmoritz @cadedaniel
